### PR TITLE
[Wip] Add API and Tests for Collectible Assemblies and AssemblyLoadContext

### PR DESCRIPTION
**Pull request by xoofx**
This is a first set of tests for collectible assemblies. This is related to adding new API for corefx #14724 and the undergoing work on the coreclr PR [Collectible Assemblies and AssemblyLoadContext](https://github.com/dotnet/coreclr/pull/8677)

This is still a WIP. I'm adding these tests for a followup with the PR on coreclr.

cc: @jkotas @gkhanna79 @kouvel @rahku

### CODE REVIEW

File: src/System.Runtime.Loader/tests/AssemblyLoadContextTests.netstandard1.7.cs [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+            return alc;
+        }
+
+        private class CollectibleChecker
+        {
+            private const int MaxGCRetry = 10;
+            private const int MaxWaitTimePerGCRetryInMilli = 10;
+
+            private readonly int expectedCount;
```
Comment (mellinoe) [inline src/System.Runtime.Loader/tests/AssemblyLoadContextTests.netstandard1.7.cs]: nit: We use a coding style where fields start with an underscore.

`expectedCount` -> `_expectedCount`
`evt` -> `_evt`
`unloadCount` -> `_unloadCount`

File: src/System.Runtime.Loader/tests/AssemblyLoadContextTests.netstandard1.7.cs [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+            private readonly ManualResetEvent evt;
+            private int unloadCount;
+
+            public CollectibleChecker(int expectedCount)
+            {
+                this.expectedCount = expectedCount;
```
Comment (mellinoe) [inline src/System.Runtime.Loader/tests/AssemblyLoadContextTests.netstandard1.7.cs]: nit: We don't use `this.` for field accesses; use `_expectedCount = expectedCount;` instead.

File: src/System.Runtime.Loader/tests/AssemblyLoadContextTests.netstandard1.7.cs (hunk 1 above)
Comment (xoofx) [inline src/System.Runtime.Loader/tests/AssemblyLoadContextTests.netstandard1.7.cs]: Fixed by commit b717442

File: src/System.Runtime.Loader/tests/AssemblyLoadContextTests.netstandard1.7.cs (hunk 2 above)
Comment (xoofx) [inline src/System.Runtime.Loader/tests/AssemblyLoadContextTests.netstandard1.7.cs]: Fixed by commit b717442

**Comment by karelz** [comment]
@xoofx I assume this will be WIP for a while (no pressure, just checking the situation ...)

**Comment by kouvel** [comment]
We'll need the API review dotnet/corefx#14724 to be approved. @rahku, could you please review dotnet/coreclr#8677? After those are completed, this PR probably would need some fixes before it'll be ready.

**Comment by karelz** [comment]
@kouvel @rahku any progress on the blockers listed above?

**Comment by karelz** [comment]
@kouvel @rahku ping?

**Comment by rahku** [comment]
This is not going to be merged soon. I plan to work on this after finishing 2.0 deliverables.

**Comment by karelz** [comment]
I'm going to close the PR then. We can reopen the PR once the dependency (https://github.com/dotnet/coreclr/pull/8677) is unblocked.

**Comment by andersborum** [comment]
What's the status on this one? AppDomains and unloading of assemblies didn't make it to the 2.0 release and I'm asking for some kind of pointer as to when we can expect this feature on .NET Core.

Thanks in advance.

**Comment by karelz** [comment]
Currently we do not have firm plans/schedule to implement the feature. There is desire to do it one day, but it is not planned yet.

**Pull request closed without merging**